lib/portage/dbapi/bintree.py

```diff
@@ -1590,6 +1736,13 @@ def getname(self, cpv, allocate_new=None):
 		except AttributeError:
 			cpv = _pkg_str(cpv)
 
+		if binpkg_format is None:
+			if hasattr(cpv, "_metadata"):
+				binpkg_format = cpv._metadata.get('BINPKG_FORMAT',
+					self.settings.get("BINPKG_FORMAT", "xpak"))
+			else:
+				binpkg_format = self.settings.get("BINPKG_FORMAT", "xpak")
+
```

Comment:
It's messy to have 3 sources of truth for binpkg_format here. We should rely on ` cpv._metadata.get('BINPKG_FORMAT', 'xpak')` alone as the source of truth, and remove the binpkg_format parameter. If that breaks something, then the caller needs to be fixed to ensure that `cpv._metadata.get('BINPKG_FORMAT', 'xpak')` returns a correct value.

Reply:
The same goes for similar methods like _allocate_filename (which getname calls). It should never use `binpkg_format = self.settings.get("BINPKG_FORMAT", "xpak")`.

Reply:
After rebase on master branch, the unit test pass with https://github.com/zmedico/portage/commit/83445643ab0c96f5928a2aa697dd59131da8f369 which does these things:

* Never use settings["BINPKG_FORMAT"] when cpv._metadata exists

* If cpv._metadata["BINPKG_FORMAT"] is empty then raise InvalidBinaryPackageFormat

* Add BINPKG_FORMAT to _pkg_str_aux_keys so that it will be automatically included in cpv._metadata["BINPKG_FORMAT"]

* Only use settings["BINPKG_FORMAT"] when allocate_new is True